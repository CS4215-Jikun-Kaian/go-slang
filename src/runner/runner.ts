import { compile } from '../compiler/compiler';
import { parse } from '../parser/parser';

export const run = async (code: string): Promise<string> => {
  return await new Promise((resolve: (value: string) => void, reject: (reason?: any) => void) => {
    try {
      const ast = parse(code);
      const instructions = compile(ast);
      // TODO: run
      resolve(
        instructions
          .map((i) => {
            const { type, ...others } = i;
            return type + (Object.keys(others).length ? ' ' + JSON.stringify(others) : '');
          })
          .join('\n')
      );
    } catch (err) {
      reject(err);
    }
  });
};
