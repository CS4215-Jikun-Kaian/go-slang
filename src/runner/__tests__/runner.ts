import fs from 'fs';
import { run } from '../runner';

const SOURCE_FILES_DIR = 'src/runner/__tests__/sourceFiles/';

describe('Runner tests', () => {
  fs.readdirSync(SOURCE_FILES_DIR).forEach((file) => {
    const source = fs.readFileSync(SOURCE_FILES_DIR + file, { encoding: 'utf8' });
    it('runs ' + file + ' correctly', async () => {
      const s = await run(source);
      expect(s).toMatchSnapshot();
    });
  });
});
