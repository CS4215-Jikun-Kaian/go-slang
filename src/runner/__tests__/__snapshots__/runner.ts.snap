// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runner tests runs simple.go correctly 1`] = `
"JOF {"addr":9}
ENTER_SCOPE {"num":0}
ENTER_SCOPE {"num":0}
LDC {"val":1}
LDC {"val":1}
BINOP {"sym":"+"}
POP
EXIT_SCOPE
EXIT_SCOPE
LDF {"arity":0,"addr":1}
ASSIGN {"pos":[0,0]}
LD {"pos":[0,0]}
CALL {"arity":0}
DONE"
`;
